src/Datadog.Trace.ClrProfiler.Managed/Integrations/Testing/AsyncTool.cs

```diff
@@ -0,0 +1,215 @@
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    return (TResult)await continuation(previousTask.Result, null, state).ConfigureAwait(false);
+                }
+
+                TResult result = default;
+                try
+                {
+                    result = await previousTask.ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    await continuation(result, ex, state).ConfigureAwait(false);
+                    throw;
+                }
+
+                return (TResult)await continuation(result, null, state).ConfigureAwait(false);
+            }
+        }
+
+        internal class TaskContinuationGenerator
+        {
+            public virtual Task SetTaskContinuation<TState>(Task previousTask, TState state, Func<object, Exception, TState, object> continuation)
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    continuation(null, null, state);
+#if NET45
+                    // "If the supplied array/enumerable contains no tasks, the returned task will immediately transition to a RanToCompletion state before it's returned to the caller."
+                    // https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?redirectedfrom=MSDN&view=netframework-4.5.2#System_Threading_Tasks_Task_WhenAll_System_Threading_Tasks_Task___
+                    return Task.WhenAll();
+#else
+                    return Task.CompletedTask;
+#endif
+                }
+
+                return InternalSetTaskContinuation(previousTask, state, continuation);
+            }
+
+            private static async Task InternalSetTaskContinuation<TState>(Task previousTask, TState state, Func<object, Exception, TState, object> continuation)
+            {
+                try
+                {
+                    await previousTask.ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    continuation(null, ex, state);
+                    throw;
+                }
+
+                continuation(null, null, state);
+            }
+
+            public virtual Task SetTaskContinuationAsync<TState>(Task previousTask, TState state, Func<object, Exception, TState, Task<object>> continuation)
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    return continuation(null, null, state);
+                }
+
+                return InternalSetTaskContinuationAsync(previousTask, state, continuation);
+            }
+
+            private async Task InternalSetTaskContinuationAsync<TState>(Task previousTask, TState state, Func<object, Exception, TState, Task<object>> continuation)
```

Comment:
nit: This could also be static but I don't know if it makes too much of a difference

Reply:
Yeah, good catch!, that `static` is missing.

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/802/commits/e0f0ca4086af7aceb3115853fbefc6f082948024